openedx/xblock-lti-consumer
# [MST-1473] Proctoring Service Launch Implementation

### Description

This pull request implements the [LTI Proctoring Services](http://www.imsglobal.org/spec/proctoring/v1p0) standard LTI messages and proctoring flow. It does not implement the Assessment Control Service.

This pull request relies on the `location` being fully decoupled from the XBlock (https://github.com/openedx/xblock-lti-consumer/issues/273) and on the ability for consuming contexts to pass relevant context to the the library. As such, this is a work-in-progress and will evolve as those requirements are met.

Because of the dependency on the XBlock, you will find a number of TODOs in this pull request. I am hoping to address these iteratively as I receive feedback.

I have also not added tests yet, to avoid churn.

### Key Details

* I added a set of views that implement the LTI Proctoring Services launch flow: `start_proctoring`, `launch_gate_endpoint_proctoring`, `start_assessment` and `end_assessment`. You'll note that `launch_gate_endpoint_proctoring` has a lot of overlap with `launch_gate_endpoint` in https://github.com/openedx/xblock-lti-consumer/pull/254. Because of the dependency on the XBlock, I could not easily reuse the logic there; I felt having a separate view for now would be easier.
* I added an `LtiProctoringConsumer`.
* I added the above views to the URL configuration.

### Dependencies
* https://github.com/openedx/xblock-lti-consumer/pull/254

### Testing Instructions

**Setup Steps:**

1. Clone the [xblock-lti-consumer](https://github.com/openedx/xblock-lti-consumer) repository into your `{workspace}/src` directory. `{workspace}` should be the directory in which you have your other edX git repositories and should be the same directory your `devstack` repository is located in.
2. Check out this branch (`mroytman/MST-1473-proctoring-service-launch`) locally.
3. Clone the [edx-exams](https://github.com/edx/edx-exams) repository into your `{workspace}` directory. `{workspace}` should be the directory in which you have your other edX git repositories and should be the same directory your `devstack` repository is located in.
4. Check out the branch `mroytman/MST-1473-proctoring-service-launch` locally.
5. Follow the [One Time Setup](https://github.com/edx/edx-exams#one-time-setup) instructions.

**Testing Instructions:**

If you want to test this implementation against the IMS Proctoring Specification testing tool, follow these steps.

1. Sign up for an IMS account [here](https://site.imsglobal.org/user/register), if you are not already a member. The approval process should take around a day.
2. You will need to set up port forwarding so that your running Django server can connect with the IMS testing tool. You can use `ngrok` for this or any other tool you like. `ngrok` has comprehensive documentation for how to accomplish this.
3. In your local checkout of `mroytman/MST-1473-proctoring-service-launch` in `edx-exams`, change `LMS_ROOT_URL` in `local.py` to the forwarded URL. For example, `https://9c11-73-238-153-173.ngrok.io`.
4. Once your membership has been approved, go to the testing tool [here](https://certification.imsglobal.org/proctoring/launch).
5. In the testing tool, click `Get Started`.
6. Enter your name, email, and organization and click `Confirm`.
7.  Click `Create` under `Create New Product`.
8. Enter a `Product` and `Version` and click `Confirm`. These values can be anything for the purposes of testing.
9. Click `Select` under `Assessment Platform`.
11. Click `Begin Configuration`.
12. Open up the Django admin by visiting `localhost:18740/admin` and logging in with the default edX development login credentials.
13. Under `LTI_CONSUMER > Lti configuration`, click `Add`.
14. Enter the following configuration:
15.1. `Version`: `LTI 1.3 (with LTI Advantage Support)`
13.2. `Config store`: `Configuration Stored on this model`.
16. From the testing tool, copy the following configuration into the `Lti configuration` model in the Django admin.
14.1. `Tool End-Point to Receive Launches` > `Lti 1p3 launch url`
14.2. `OIDC Login Initiation URL` > `Lti 1p3 oidc url`
17. Click `Next`.
18. Enter the following configuration in the testing tool. In this example, `{root}` refers to the URL that is being forwarded. For example, `https://9c11-73-238-153-173.ngrok.io`.
16.1. `Testing iss Value`: `{root}`
16.2. `OIDC Auth URL`: `{root}/lti/launch_proctoring`
16.3. `Platform Well-Known/JWKS URL`: `{root}/lti/public_keyset`
16.4. `Client Id (Assigned to Cert Suite)` > Enter the `Lti 1p3 client id` from the `Lti configuration` model in the Django admin.
19. Click `Next`.
20. Enter `1` for the deployment ID and click `Next`.
21. From the testing tool, copy the`Proctoring Tool JWKS URL` into the `Lti configuration` model in the Django admin in the `Lti 1p3 tool keyset url` field.
22. Click `Next`.
23. Check `Pause on OIDC Exchanges to Inspect Values` and/or `Pause Prior to JWT Messages to Inspect Values` depending on whether you want interstitials before Tool requests or responses to the Assessment Platform.
24. Visit `{root}/api/start_proctoring` to kick off the flow.
